### PR TITLE
Refactor object properties to become real fields.

### PR DESCRIPTION
**Pull request by zherczeg**
The positive side effect is that the maximum Jerry memory is increased to 512K.
Furthermore a slight (1.3%) performance improvement was measured on RPi2 with
SunSpider.

JerryScript-DCO-1.0-Signed-off-by: Zoltan Herczeg zherczeg.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by zherczeg** [comment]
Review please.

File: jerry-core/ecma/base/ecma-globals.h [hunk 1]
```diff
@@ -279,151 +279,95 @@ typedef struct ecma_property_t
...
+ * Description of ECMA-object or lexical environment
+ * (depending on is_lexical_environment).
  */
-#define ECMA_OBJECT_LEX_ENV_OUTER_REFERENCE_CP_POS (ECMA_OBJECT_LEX_ENV_TYPE_POS + \
-                                                    ECMA_OBJECT_LEX_ENV_TYPE_WIDTH)
-#define ECMA_OBJECT_LEX_ENV_OUTER_REFERENCE_CP_WIDTH (ECMA_POINTER_FIELD_WIDTH)
+typedef struct ecma_object_t
+{
+  /** object type, flags, and refs */
+  uint16_t type_flags_refs;
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-globals.h]: Could you add a more descriptive comment for this? For example it is not trivial why 
`if (object_p->type_flags_refs < ECMA_OBJECT_MAX_REF)` will work. (I know it is correct at the moment :) ) I think we need to give some help here. How should be tpye/flags/refs used?

**Comment by zherczeg** [comment]
Fixed.

File: jerry-core/ecma/base/ecma-gc.c [hunk 2]
```diff
@@ -180,16 +137,20 @@ ecma_gc_set_object_visited (ecma_object_t *object_p, /**< object */
     is_visited = !is_visited;
   }
 
-  object_p->container = JRT_SET_BIT_FIELD_VALUE (uint64_t, object_p->container,
-                                                 is_visited,
-                                                 ECMA_OBJECT_GC_VISITED_POS,
-                                                 ECMA_OBJECT_GC_VISITED_WIDTH);
+  if (is_visited)
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-gc.c]: Aren't the above 5 lines simply equal to `if (is_visited != ecma_gc_visited_flip_flag)`?

File: jerry-core/ecma/base/ecma-gc.c [hunk 3]
```diff
@@ -88,88 +88,45 @@ static void ecma_gc_mark (ecma_object_t *object_p);
...
-                                                 refs,
-                                                 ECMA_OBJECT_GC_REFS_POS,
-                                                 ECMA_OBJECT_GC_REFS_WIDTH);
-} /* ecma_gc_set_object_refs */
-
-/**
  * Get next object in list of objects with same generation.
  */
-static ecma_object_t *
+static inline ecma_object_t *
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-gc.c]: Are these (`ecma_gc_{get,set}_object_next`) functions still needed? With the proposed changes, they are merely wrappers around `ECMA_{GET,SET}_POINTER`. (`ecma_gc_{get,set}_object_refs` have been dropped as well, not caring about the asserts. So, do we need these?)

Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: To me this style follows Jerry logic better. But I can remove them if you wish.

Comment (akiss77) [inline jerry-core/ecma/base/ecma-gc.c]: I'm only looking for consistency. Why have `ecma_gc_{get,set}_object_refs` been removed then?

I thought that all these getters/setters were used/useful to access multiple data packed together in a 64-bit container bitfield. As that container is now being separated into multiple fields that have their own names, accessing them through getters/setters is unnecessary. At least, that was my impression.

Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: That is a different reason. Getting and setting is used for incrementing and decrementing reference counters, but that is an inefficient way to do that (lots of shifting, masking). After the refcounter updates are optimized, those functions are not used anymore, so they are removed.

Comment (akiss77) [inline jerry-core/ecma/base/ecma-gc.c]: Thanks for the explanation. OK with me then.

**Comment by akiss77** [comment]
LGTM (FWIW)

**Comment by LaszloLango** [comment]
@zherczeg, thanks for the update. LGTM

**Comment by zherczeg** [comment]
Landed in d674b92f266c4bd8892d823e041a8d8f1f77a961

**Pull request closed without merging**